### DISCUSSION TITLE
Undefined index error when emitting event which no longer exists

### DISCUSSION THREAD

**Discussion by mjerwin** [livewire/livewire#3116] [7. Bugs (archived)]
### Description
When using `getListeners()` to define dynamic listeners if one of those dynamic listeners is removed then you get an undefined index error.

**I've got a fix for this ready to PR if this deemed to be a valid issue.**
Suggested fix is to check for undefined indexes in `\Livewire\ComponentConcerns\ReceivesEvents::fireEvent`

### Exact steps to reproduce

1. Override `getListeners()` to add dynamic events
2. Call an action which would cause one of the dynamic events to be removed
3. Emit the event which has now been removed
4. See "Undefined index" error

### Stripped-down, copy-pastable code snippets
```php
class ItemList extends Component
{
    public $items = [];

    public function mount()
    {
        $this->items = [
            1 => ['name' => 'Item 1'],
            2 => ['name' => 'Item 2'],
            3 => ['name' => 'Item 3'],
        ];
    }

    public function render()
    {
        return view('livewire.item-list');
    }

    public function delete($id)
    {
        unset($this->items[$id]);
    }

    public function getListeners()
    {
        return collect($this->items)
            ->mapWithKeys(function ($item, $key) {
                return ['itemUpdated:' . $key => '$refresh'];
            })
            ->toArray();
    }
}
```

```javascript
Livewire.emit('itemUpdated:2');
```

```
ErrorException: Undefined index: itemUpdated:2

/Users/mje/Sites/packages/livewire/src/ComponentConcerns/ReceivesEvents.php:72
/Users/mje/Sites/packages/livewire/src/HydrationMiddleware/PerformEventEmissions.php:19
/Users/mje/Sites/packages/livewire/src/LifecycleManager.php:70
/Users/mje/Sites/packages/livewire/src/Connection/ConnectionHandler.php:12
/Users/mje/Sites/packages/livewire/src/Controllers/HttpConnectionHandler.php:13
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php:48
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Route.php:254
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Route.php:197
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Router.php:692
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Router.php:694
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Router.php:669
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Router.php:635
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Routing/Router.php:624
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:166
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:141
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:110
/Users/mje/Sites/packages/livewire/vendor/laravel/framework/src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php:508
/Users/mje/Sites/packages/livewire/src/Testing/TestableLivewire.php:184
/Users/mje/Sites/packages/livewire/src/Testing/MakesHttpRequestsWrapper.php:29
/Users/mje/Sites/packages/livewire/src/Testing/TestableLivewire.php:185
/Users/mje/Sites/packages/livewire/src/Testing/TestableLivewire.php:173
/Users/mje/Sites/packages/livewire/src/Testing/Concerns/MakesCallsToComponent.php:134
/Users/mje/Sites/packages/livewire/src/Testing/Concerns/MakesCallsToComponent.php:22
/Users/mje/Sites/packages/livewire/src/Testing/Concerns/MakesCallsToComponent.php:15
/Users/mje/Sites/packages/livewire/tests/Unit/ComponentEventsTest.php:47
```

**Context**
- Livewire version: 2.3.8
- Laravel version: N/A
- Alpine version: N/A
- Browser: Chrome

**Comment by joshhanley** [comment]
@mjerwin I'd suggest submitting your PR with a test/tests that demonstrate the issue and the fix so others can review it.

Hope this helps!

**Comment by mjerwin** [comment]
Hey, @joshhanley. Thanks! I was a little hesitant to submit the PR before the issue had been verified but to be fair I'm pretty sure it's an issue others would come across not just me so have submitted it.

#2427

**Comment by joshhanley** [comment]
@mjerwin no worries! Thanks for submitting, I have added my comments to it 🙂

**Comment by adamhopkinson** [comment]
I came across this big yesterday, but it was my first time with Livewire listeners so at the time I put it down to my own mistakes.

+1 for getting this PR in as soon as!

**Comment by Jnesselr** [comment]
Definitely an issue others will come across, as I just came across it now :)

I don't know if this is out of scope (I don't think it is), but can we also get livewire to correctly unsubscribe from channels that are no longer needed? ie, I have a channel that's basically `private-jobs.foo` and depending on the lifecycle of the component, it might subscribe to `private-jobs.bar`. What I expect to see is that `private-jobs.foo` gets unsubscribed and `private-jobs.bar` gets subscribed but the first stays subscribed and the second doesn't. That PR would fix the lookup of the event but honestly even that's avoidable if it unsubscribes from channels it no longer needs.

To fix that, it looks like we'd have to have the listeners be updated every time in hydrate, not just in the initial section: https://github.com/livewire/livewire/blob/master/src/RenameMe/SupportEvents.php#L18
